### PR TITLE
Bluetooth: Controller: DTM: Fix timer rollover

### LINKED ISSUE #42534: BLE Testing functions do not work properly
I am using Nordic nRF5340DK with Zephyr 2.7.0 LTS SDK. I am using built in j-link and SES(Segger Embedded Studio) to debug. I also use emptyapp from nordic on the application core and example HCI uart from /samples/bluetooth/hci_uart and set CONFIG_BT_CTLR_DTM_HCI=y in prj.conf

I am using nrf testing applications like RSSI Viewer and DTM test demos on Windows and code for nRF5340DK is being build in linux Debian enviroment. 

**The issue:**
When I use HCI command LE Transmitter Test the test packets stop transmitting after certain time or packet count(around ~4000).
I cant find any workaround to this issue, yet it is needed for Bluetooth hardware testing. 
The DTM sample app from nRF Connect SDK works as expected but they are not compatible with Bluetooth stack.

**To Reproduce**
Build an Empty Application Core from Nordic, to give control for network core, from their repository. 
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/nrf5340/empty_app_core/README.html

Build an HCI-UART application with setting:  **CONFIG_BT_CTLR_DTM_HCI=y** 

1.Run SES and load HCI-UART example application to the network core,
2.load empty application core to the network core,
3.connect the UART of nrf5340 to rtt-uart output of NRF5340DK to proceed with attaching HCI-BLE to linux
4. On linux(debian in my case) run as su : _btattach -B /dev/ttyACM1 -S 1000000 -P h4 &_
5. now open the RSSI Viewer application on another Nordic developement kit and start it to watch the BLE radio spectrum.
6. now on the linux run the hci tool and send command: _hcitool -i hci1 cmd 0x08 0x1e 0x02 0x20 0x00 0x00_ (HCI_LE_Transmitter_Test 
[v2] from specification BLE 5.2 Vol 4, Part E) It should start sending packets with PRBS9 at channel 2 with length of 32 bytes, until issuance of the command hcitool -i hci1 cmd 0x08 0x1f (LE_Test_End)

**Expected behavior**
The Bluetooth stack should transmit test packets continuously until receiving LE_Test_End command from HCI interface, yet it spontaneously stops transmitting sooner than the command LE_Test_End.

**Additional info**
The same issue happens when in network core app directly is being called ll_test_tx() function from subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c - the packets suddenly stop transmitting after some time.

### PR DESCRIPTION
**Pull request by piotrParf**
It changes the behaviour of while loop in lll_test.c in isr_tx() . When bug happens and t<s is quite big it can hapen that while loop will take a long time. This leads to setting up start of a radio timer before current timer value. This then leads to invocation of a timer after rollout and this will lead to very large delay in sending packet. 

Fixes #42534.

Signed-off-by: Piotr Parfeniuk [piotr.parfeniuk@magitech.pl](mailto:piotr.parfeniuk@magitech.pl)

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c [hunk 1]
```diff
@@ -113,8 +113,9 @@ static void isr_tx(void *param)
 	/* Set timer capture in the future. */
 	radio_tmr_sample();
 	s = radio_tmr_sample_get();
-	while (t < s) {
-		t += SCAN_INT_UNIT_US;
+	/*Previous calculation in while loop can take too long time and can rollout the timer*/
+	if (t < s){
+		t+=((((s - t) / SCAN_INT_UNIT_US) + 1) * SCAN_INT_UNIT_US);
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:118]: Please add roll over handling. Feel free to change the implementation for better readability.

```suggestion
	if (t < s) {
		t += ((((s - t) / SCAN_INT_UNIT_US) + 1) * SCAN_INT_UNIT_US);
	} else {
		/* Handle rollover case here, timer is configured to 24-bits */
		t += ((((BIT(24) - t + s) / SCAN_INT_UNIT_US) + 1) * SCAN_INT_UNIT_US);
```

**Review by cvinayak — CHANGES_REQUESTED** [review]

**Comment by carlescufi** [comment]
@piotrParf ping, can you take a look at the review comments please?

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c [hunk 2]
```diff
@@ -105,29 +105,32 @@ static void isr_tx(void *param)
...
-	s = radio_tmr_sample_get();
-	while (t < s) {
-		t += SCAN_INT_UNIT_US;
+	sampled_time = radio_tmr_sample_get();
+	/*Previous calculation in while loop can take too long time and can rollout the timer*/
+	if (transmit_time < sampled_time){
+		transmit_time+=((((sampled_time - transmit_time) / SCAN_INT_UNIT_US) + 1) * SCAN_INT_UNIT_US);
 	}
+	/* Handle rollover case here, timer is configured to 24-bits */
+	transmit_time=transmit_time&0xffffff;
```
[the remark this answers is not in the document]
Comment (piotrParf) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:121]: Ok. Can you explain me if the difference will be too small or sth ?

[the remark this answers is not in the document]
Comment (piotrParf) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:121]: > How is `transmit_time = 0xffffff` and `sampled_time = 0x000001` handled in this?

@[cvinayak] can you respond and explain me the problem better ?

[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:121]: `sampled_time = radio_tmr_sample_get();`  when gets a value after rollover and `transmit_time` is a value before rollover then `transmit_time` will be greater than `sampled_time`, you changes does not handle rollover

[the remark this answers is not in the document]
Comment (piotrParf) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c:121]: Ahh I see it. There was also another issue causing similar issue at nrf5340 in 2.7 with default setting of CONFIG_BT_CTLR_TIFS_HW=n (to change it, it is needed to also turn on BT_CTLR_ADVANCED_FEATURES and disable BT_CTLR_RADIO_ENABLE_FAST). And this fails working at packet radio timer ending time in 2.7.0 by reading always 0 with radio_tmr_end_get. It is fixed in never versions of zephyr(checked 3.0.0) in function radio_tmr_end_capture() and the change of changing defines is fixing the issue with CONFIG_BT_CTLR_TIFS_HW=n.

File: subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c (hunk 1 above)
Comment (piotrParf) [reply subsys/bluetooth/controller/ll_sw/nordic/lll/lll_test.c]: should be done

**Comment by piotrParf** [comment]
Why do i got .github/labeler.yml Error: HttpError ?

**Comment by carlescufi** [comment]
@KAGA164 could you please comment on this proposed change?

**Comment by KAGA164** [comment]
> @KAGA164 could you please comment on this proposed change?

@carlescufi I will check timings with Anritsu

**Comment by KAGA164** [comment]
@piotrParf  Please rebase and align your PR due to changes made by #42300. I included CTE time in the transmit time calculation.

**Review by KAGA164 — CHANGES_REQUESTED** [review]
The DTM Tx interval is incorrect with this fix because it doesn't follow the calculation from Core v5.3 Vol 6 Part F 4.1.6 where interval should follow basically pattern `n*625us`. I did a test on 1Mb PHY with payload length 37 bytes and I measured the 908us interval when it shall be 625us.
DTM_interval

**Comment by piotrParf** [comment]
> The DTM Tx interval is incorrect with this fix because it doesn't follow the calculation from Core v5.3 Vol 6 Part F 4.1.6 where interval should follow basically pattern `n*625us`. I did a test on 1Mb PHY with payload length 37 bytes and I measured the 908us interval when it shall be 625us. DTM_interval

As I do not have proper tools for this what is your suggestion ?

**Comment by KAGA164** [comment]
@piotrParf I made PR with my proposal how to handle Tx timings in the DTM, Could you test it and give me a feedback ? https://github.com/zephyrproject-rtos/zephyr/pull/44735

**Comment by piotrParf** [comment]
> @piotrParf I made PR with my proposal how to handle Tx timings in the DTM, Could you test it and give me a feedback ? #44735

For me it is many changes to 2.7.0 but it looks like it works when loaded with this controller to network core.

**Pull request closed without merging**